feature/oracle_profiler

lolo115 opened the pull request:
## Changes
Creating the oracle extractor.
Unit tests passed for:
- dialog to configure the extractor
- run scripts to gather data
- local database population with extracted data from the Oracle Database.
### What does this PR do?
Oracle profiler data extraction
### Relevant implementation details
The data analysis is still WIP.

### Caveats/things to watch out for when reviewing:

### Linked issues
<!-- DOC: Link issue with a keyword: close, closes, closed, fix, fixes, fixed, resolve, resolves, resolved. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword -->

Resolves #..

### Functionality

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs lakebridge ...`
- [ ] ... +add your own

### Tests
<!-- How is this tested? Please see the checklist below and also describe any other relevant tests -->

- [x] manually tested
- [x] added unit tests
- [ ] added integration tests

m-abulazm commented on src/databricks/labs/lakebridge/connections/database_manager.py:
```diff
@@ -152,6 +169,8 @@ def fetch(self, query: str) -> FetchResult:
 
     def check_connection(self) -> bool:
         query = "SELECT 101 AS test_column"
+        if self._db_type.lower() == "oracle":
```
move this method to the connectors

m-abulazm replied:
https://github.com/databrickslabs/lakebridge/pull/2421

m-abulazm commented on tests/integration/assessments/test_profiler.py:
```diff
@@ -8,6 +8,37 @@
 
 
+# Tests for Oracle profiler section
+def test_Oracle_as_supported_source_technologies() -> None:
```
missing tests

m-abulazm commented on src/databricks/labs/lakebridge/resources/assessments/validation/oracle_extract_schema.yml:
```diff
@@ -0,0 +1,138 @@
+source_tech: oracle
```
investigate usage

m-abulazm commented on tests/test_blueprint_oracle.py:
```diff
@@ -0,0 +1,11 @@
+profiler = Profiler("oracle", None)
+# supported_platforms = profiler.supported_platforms()
+# print(supported_platforms)
+profiler.profile(pipeline_config=config)
```
clean

m-abulazm commented on tests/resources/assessments/pipeline_oracle_config_main.yml:
```diff
@@ -0,0 +1,11 @@
+name: oracle_assessment
```
investigate

m-abulazm reviewed (changes requested):
missing docs

dgomez04 commented on uv.lock:
Can we check with @asnare if there are changes required here? I think we shouldn't be updating `uv.lock` based on some work on the CI.

dgomez04 commented on pyproject.toml:
```diff
@@ -41,6 +41,7 @@ dependencies = [
   "SQLAlchemy~=2.0.40",
   "pygls~=2.1.1",
   "duckdb>=1.2.2,<1.5.0",
+   "oracledb==3.4.0",
```
The **exact pin** is inconsistent with the rest of the file. Every other dependency uses `~=` or `>=,<` ranges.

dgomez04 commented on .gitignore:
Are these needed here? Weren't they being ignored beforehand?

dgomez04 commented on src/databricks/labs/lakebridge/assessments/configure_assessment.py:
```diff
@@ -72,6 +73,43 @@ def run(self):
+            "secret_vault_name": secret_vault_name,
+            source: {
+                "host": self.prompts.question("Enter the host details (Server name, IP address, SCAN Name)"),
+                "tnsPort": int(
```
`tnsPort` and `tnsService` uses camelCase where the existing configurators use snake_case. This is incosnistent and we should standardize.

dgomez04 commented on src/databricks/labs/lakebridge/assessments/configure_assessment.py:
```diff
@@ -72,6 +73,43 @@ def run(self):
+                "tnsService": self.prompts.question(
+                    "Enter the TNS service name as registered in the Oracle listener", default="orcl"
+                ),
+                "user": self.prompts.question("Enter user name with system privileges", default="SYSTEM"),
```
The documentation tells users to create a common profiler user `c##lakebridge_profiler` but the default prompt undermines this and defaults to `SYSTEM` - a priviledged Oracle account. Default should either empty or `c##lakebridge_profiler`.

m-abulazm replied:
I would still default to SYSTEM. 
the docs explain what to do for least privileged access. or we can remove the default completely

dgomez04 replied:
Honestly - I'd remove the default to make sure users are completely aware of what they're selecting (considering they'll need to select a password afterwards). Let me know if you agree.

dgomez04 commented on src/databricks/labs/lakebridge/assessments/configure_assessment.py:
```diff
@@ -72,6 +73,43 @@ def run(self):
+        secret_vault_type = str(self.prompts.choice("Enter secret vault type (local | env)", ["local", "env"])).lower()
+        secret_vault_name = None
+
+        logger.info("Please refer to the documentation to understand the difference between local and env.")
```
This is redundant, lines 83-86 explain this already.

dgomez04 commented on src/databricks/labs/lakebridge/assessments/configure_assessment.py:
```diff
@@ -72,6 +73,43 @@ def run(self):
+        credential = {
+            "secret_vault_type": secret_vault_type,
+            "secret_vault_name": secret_vault_name,
+            source: {
```
The `credential` dict here jumps straight from `secret_vault_*` keys into Oracle connection parameters with no `auth_type` field. All other configurations set `auth_type` at the top of the `source` block. This matters because Oracle's enterprise authentication surface is meaningfully richer than SQL Server's, and several of these might be common asks for the actual customers we're targeting: 
- **Oracle Wallet** (TNS_ADMIN + sqlnet.ora + cwallet.sso) — stored encrypted credentials, very common in regulated environments where security teams ban plaintext DB passwords in YAML files.
- **External / Kerberos / OS authentication** — for SSO and AD integration.
- **Token-based auth for Oracle Cloud** (OCI IAM tokens) and Autonomous Database.
- **mTLS / TCPS with client certs** — required for Autonomous Database "TLS-only" mode.

No need to actually implement these types today - just reserve the field and document the one currently-supportede value, like `MSSQLConnector`does.

m-abulazm replied:
we can do a followup to support different auth methods

dgomez04 commented on src/databricks/labs/lakebridge/connections/database_manager.py:
```diff
@@ -79,7 +79,7 @@ def _create_connector(db_type: str, config: JsonObject) -> DatabaseConnector:
         "snowflake": SnowflakeConnector,
         "mssql": MSSQLConnector,
         "tsql": MSSQLConnector,
-        "synapse": MSSQLConnector,  # Synapse uses MSSQL protocol
```
Why is `synapse` deleted here? Is `snowflake` already implemented?

dgomez04 commented on src/databricks/labs/lakebridge/connections/database_manager.py:
```diff
@@ -125,9 +125,26 @@ def _connect(self) -> Engine:
         return create_engine(connection_string)
 
 
+class OracleConnector(_BaseConnector):
```
- No support for other authentication methods, such as TCPS (TLS), Oracle Wallet, or RAC easy-connect strings beyond host:port/service.
- Does not pass `mode=oracledb.SYSDBA` etc., which is irrelevant for SELECT-only profilers but worth noting.
- No `loginTimeout` equivalent (the MSSQL connector at least sets `"loginTimeout": "30"`).

m-abulazm replied:
lets iterate on the connection in a followup

dgomez04 commented on src/databricks/labs/lakebridge/connections/database_manager.py:
```diff
@@ -152,6 +169,8 @@ def fetch(self, query: str) -> FetchResult:
 
     def check_connection(self) -> bool:
```
`DatabaseManager` is supposed to be the engine-agnostic layer that dispatches to a connector via `_create_connector`; once that dispatch happens, no code in `DatabaseManager` should be re-deriving the engine via string checks on `db_type`.

**Concrete suggestion:** have each connector own its own smoke-test query as a class attribute, and let `DatabaseManager` just read it.

```python
class _BaseConnector(DatabaseConnector):
    # Subclasses override this with a dialect-appropriate one-row SELECT.
    smoke_test_query: str = "SELECT 101 AS test_column"
    ...

class OracleConnector(_BaseConnector):
    smoke_test_query = "SELECT 101 AS test_column FROM dual"
    ...

class DatabaseManager:
    def __init__(self, db_type: str, config: JsonObject):
        self.connector = _create_connector(db_type, config)
        # self._db_type = db_type   ← can be deleted

    def check_connection(self) -> bool:
        result = self.fetch(self.connector.smoke_test_query)
        if result is None:
            return False
        return result.rows[0][0] == 101
```

m-abulazm replied:
this is fixed on https://github.com/databrickslabs/lakebridge/pull/2421. please review

dgomez04 replied:
When this PR rebases on that, please confirm: (a) the db_type == "oracle" branch in check_connection is gone, (b) OracleConnector.health_check() returns SELECT 101 FROM dual, (c) self._db_type attribute is removed from __init__, and (d) a unit test covers the new OracleConnector.health_check() path.

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/pipeline_config.yml:
```diff
@@ -0,0 +1,136 @@
+    mode: overwrite
+    frequency: once
+    flag: active
+  - name: config_containers
```
Every logical extractor here is split into two YAML steps that share the same name: — one with `type: ddl`, one with `type: sql`. That works today because `pipeline.py` happens to dispatch on `step.type` first and uses `step.name` as the destination DuckDB table name in both branches.

But the duplicate names cause the following problems: 
- `PipelineClass.execute()` aggregates `StepExecutionResult` objects keyed by `step.name`, then logs `failed_steps = [r for r in execution_results if r.status == StepExecutionStatus.ERROR]`. If the SQL half of `config_containers` fails, the user sees Pipeline execution failed due to errors in `steps: config_containers` with no way to tell whether the DDL or the SQL step failed. This makes debugging a headache.
- The DDL step does if not `self._table_exists(conn, step.name): conn.execute(ddl)` — i.e., it implicitly assumes the `CREATE TABLE` in the DDL file uses a table named exactly `step.name`. There's nothing enforcing that; if a future edit changes the name in one place but not the other, the pipeline silently no-ops the DDL or fails the insert.
- Anyone refactoring `pipeline.py` to dedupe steps by name (perfectly reasonable assumption) breaks this YAML silently.

I think we need to discuss how to handle it, one alternative could be giving DDL steps a different suffix:
```yaml
- name: config_containers_schema
  type: ddl
  extract_source: .../config_containers_ddl.sql
- name: config_containers
  type: sql
  extract_source: .../config_containers.sql
```

The `_execute_ddl_step` check would still need to know the target table name (which is currently inferred from `step.name`); the cleanest version of that is to either parse the table name out of the CREATE statement, or add a `target_table: field` to Step so name (the identifier in logs/errors) and target (the DuckDB table the step writes to) are decoupled. The second is a small change in `profiler_config.py` and `pipeline.py` and is worth doing now while the file count is small.

**Longer-term thought:** it might be cleaner to fold DDL into the SQL step entirely — i.e., a single step with `setup_ddl: ...` and `extract_sql: ... fields`, since DDL+SQL are conceptually one extractor — but that's a bigger change and probably belongs in a follow-up. **We should probably file an issue for this.**

m-abulazm replied:
Agreed. lets have a followup for that

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/pipeline_config.yml:
```diff
@@ -0,0 +1,136 @@
+name: oracle_assessment
+version: "1.0"
+extract_folder: "/tmp/data/oracle_assessment"
```
Three things bundled here:
- `/tmp/...` isn't a real Windows path — `Path("/tmp/...").expanduser()` in `pipeline.py:42` resolves it to `<current drive>:\tmp\data\oracle_assessment`, which "works" but isn't where any Windows user expects an extract to land. The Synapse profiler already solves this with `~/.databricks/labs/lakebridge_profilers/synapse_assessment`, which `expanduser()` resolves correctly on every OS.
- Two of the three existing profilers (`mssql`, this `oracle` one) write to `/tmp/data/...`, while `synapse` writes to `~/.databricks/labs/lakebridge_profilers/...`. A user running multiple profilers ends up needing two different `--extract-file` paths in their `create-profiler-dashboard` command for no good reason.
- macOS purges files in `/tmp` periodically; AWR-based extracts that take 30+ minutes to produce shouldn't live there.

Could we standardize on the Synapse-style path here, e.g.
```bash
extract_folder: "~/.databricks/labs/lakebridge_profilers/oracle_assessment"
```
_If that's done, the integration test (tests/integration/assessments/test_profiler.py) needs its hardcoded Path("/tmp/data/oracle_assessment/profiler_extract.db") assertion updated to match._

m-abulazm replied:
lets clean this up in a separate PR and investigate how to standardize it

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/pipeline_config.yml:
Two pairs of files exist under `oracle/` that are not referenced anywhere in this `pipeline_config.yml`:
- `test_nulloutput.sql` + `test_nulloutput_ddl.sql` — a `SELECT 'value' as X from dual where 1=2` query and a matching `CREATE TABLE`. Looks like author-local debugging that wasn't cleaned up before opening the pull request. Delete.
- `perf_heatmap_raw.sql` — this one is more concerning. It's a SQL*Plus driver script:
```sql
declare
  cursor c is select inst_id from gv$instance order by 1;
begin
  for i in c loop
    dbms_output.put_line('@./scripts/main/perf_addon/perf_hm_raw.sql '||i.inst_id);
    ...
  end loop;
end;
/
spool off
start ./scripts/main/perf_addon/script.sql
exit
```

A few problems with this file:
- The `declare ... begin ... end`; `/`, `spool off`, `start`, `exit` directives are SQL*Plus-specific. None of them are valid through SQLAlchemy / oracledb — they'd just error.
- It references `./scripts/main/perf_addon/perf_hm_raw.sql` and `./scripts/main/perf_addon/script.sql`, neither of which exists in the pull request.
- `dbms_output.put_line` would silently do nothing without `dbms_output.enable()` and an output handler attached on the client side.
- Even if all of the above were fixed, the file isn't referenced in `pipeline_config.yml`, so it would never be invoked.

**Please review these.**

m-abulazm replied:
cleaned up

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/config_db_features.sql:
Three nits in one file:
- **Stray commented-out alternative** on line 6 (-- `select listagg(distinct name,',')...`) sitting inside a `WITH` clause. Either delete it or, if it's the "right" version, swap it in. Leaving it there as an out-of-tree alternative will confuse future readers.
- **Implicit-join syntax** in the inner subqueries (`FROM gv$osstat WHERE stat_name in (...)` etc.) and ANSI-style UNION between heterogeneous shapes is hard to read. The query produces six column shapes via null as ... placeholders to make them union-compatible — that's clever but signals the data is really three different result sets glued together. Consider returning them as separate steps (`config_db_version`, `config_db_pdbs`, `config_db_cpu`) — it's a nicer separation, and downstream dashboarding gets cleaner table shapes.
- **Implicit join in the outer query** could become explicit `JOIN` syntax. Same correctness, much easier to grep when something breaks.

m-abulazm replied:
done

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/config_storage.sql:
There are two commented-out branches that would have classified TEMP tablespaces (lines 5 and ~27 in the case and the matching group by case). Either:
- **Implement it:** uncomment the branch in both places, add 'TEMP' to the doc's tablespace-type list, and confirm `cdb_data_files` actually contains TEMP rows on your target Oracle versions (TEMP files live in `dba_temp_files` / `cdb_temp_files`, not `cdb_data_files`, so the commented-out logic might not even work as written).
- **Delete it:** drop both `-- when ...` lines and document that TEMP tablespaces are out of scope for the storage extract.

m-abulazm replied:
done

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/config_pdb_objects.sql:
No change needed in this file — the `oracle_maintained='N'` filter against cdb_users is the right way to exclude Oracle-internal schemas on 12.2+. Calling it out only because `perf_fgd_session_evolution.sql` does the same job by hardcoding a list of ~30 schema names; the two should be consistent. See my comment on that file.

m-abulazm replied:
done

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/config_pdb_partitions.sql:
No change needed in this file — the `oracle_maintained='N'` filter against `cdb_users` is the right way to exclude Oracle-internal schemas on 12.2+. Calling it out only because `perf_fgd_session_evolution.sql` does the same job by hardcoding a list of ~30 schema names; the two should be consistent. See my comment on that file.

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/perf_cpu_waits.sql:
The inner subquery `TO_CHARs` sample_time to `'YYYY-MM-DD HH24'` (so it's a string truncated to the hour), and the outer query then `TO_DATEs` that string back to a timestamp. Two issues:
1. **Information loss.** `sample_time` from `cdb_hist_active_sess_history` is a high-precision timestamp; round-tripping through an hour-grained string drops minute/second/fractional precision unrecoverably. If you want hourly aggregates, that's fine — but please use `TRUNC(sample_time, 'HH24')` directly, which keeps it as a `DATE` and is also faster (no string conversion).
2. **DDL/data mismatch.** `perf_cpu_waits_ddl.sql` declares `mtime TIMESTAMP`, which suggests sub-second fidelity. After the round-trip, `mtime` only carries hour-grained values. Either change the DDL to `mtime DATE` (more honest) or change the SQL to keep the `original sample_time` precision.
3. **NLS dependency.** `TO_DATE(string, format)` is sensitive to `NLS_DATE_LANGUAGE / NLS_DATE_FORMAT` — one customer's session settings can break this. Direct `TRUNC(sample_time, 'HH24')` avoids that entirely.

Suggested rewrite:
```sql
SELECT
    instance_number,
    con_id,
    CON_DBID,
    TRUNC(sample_time, 'HH24') AS mtime,
    NVL(event, 'ON CPU') AS event,
    NVL(wait_class, 'ON CPU') AS wait_class,
    COUNT(*) * 10 AS total_wait_time
FROM cdb_hist_active_sess_history
GROUP BY instance_number, con_id, CON_DBID, TRUNC(sample_time, 'HH24'), event, wait_class
```

_Then keep `mtime DATE` (or `TIMESTAMP` if you intentionally want sub-hour granularity, in which case drop the `TRUNC`)._

m-abulazm replied:
done

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/perf_fgd_session_evolution.sql:
1. **Copy-paste bug.** The `WHERE username NOT IN (...)` clause lists the same ~30 Oracle system schemas twice within the same parenthesised list — `'SYS','SYSTEM','XS$NULL', ..., 'SYSDG'`then immediately again. Functionally harmless (Oracle dedupes IN-list literals), but a clear signal the file wasn't proofread, and it makes future edits dangerous: if someone adds a new schema to one half but not the other, future maintainers will see asymmetry and waste time figuring out whether it's intentional.
2. **The whole approach should match `cdb_objects` / `cdb_users` filtering.** The earlier `config_pdb_objects.sql` and `config_pdb_partitions.sql` queries use the much cleaner `WHERE oracle_maintained='N'` against `cdb_users`. That handles every Oracle-shipped schema correctly across versions and is one extra column instead of a 30-line hardcoded literal.

Suggested rewrite:
```sql
... ,
(
  SELECT user_id, username
  FROM cdb_users
  WHERE oracle_maintained = 'N'
) u
```
_Same shape, ~28 fewer lines, robust to Oracle adding new internal schemas in future versions._

m-abulazm replied:
done

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/perf_heatmap.sql:
The active query ends at ~line 38; the remaining ~87 lines are commented out. Two distinct dead blocks:
1. **SQL*Plus column-formatting directives** (`-- col "00-01_ " for 90.99...`) — only useful when running against `sqlplus` interactively, irrelevant under `SQLAlchemy`/`oracledb`.
2. A complete alternative implementation using 1WITH t AS (...) SELECT ... FROM t pivot(SUM(value) ...)` — this is the "wide-format" version of the heatmap.

_Both should be deleted. If the `PIVOT` version is the intended endgame (looks like it's the better shape for downstream dashboarding), please file a follow-up issue capturing the design decision and delete the commented block here. Shipping it commented-out makes it look like a feature flag and will create confusion the moment someone goes to "tweak the heatmap query."_

m-abulazm replied:
cleaned

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/perf_heatmap_raw.sql:
This file is a **SQL*Plus** driver script and won't run via SQLAlchemy/oracledb:
- `declare ... begin ... end; /`, `spool off`, `start ./scripts/...`, and exit are SQL*Plus client commands, not server-side SQL. The pipeline runner would error on the very first statement.
- It references `./scripts/main/perf_addon/perf_hm_raw.sql` and `./scripts/main/perf_addon/script.sql`, neither of which exists in the repo.
- `dbms_output.put_line` would silently produce nothing without `dbms_output.enable()` and a client-side output handler.
- It isn't referenced in `pipeline_config.yml`, so even if it worked, nothing would invoke it.

_Looks like leftover from a prior manual SQL*Plus assessment workflow that got committed by mistake. Please delete the file. If "raw heatmap" was intended to be a real feature, please describe it's intent on a follow-up issue. We shouldn't ship it on this state._

m-abulazm replied:
cleaned

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/perf_sqltext.sql:
**1. 2KB SQL sampling.** Every `dbms_lob.substr(t.SQL_TEXT, 2000)` call truncates the SQL text to the first 2KB before doing the pattern match. That's reasonable — long stored procedures aren't classified usefully — but it means the extract intentionally does not exfiltrate full SQL text. The user-facing doc (`oracle.mdx`) is currently silent on this; please add a one-liner there saying "SQL text is sampled to the first 2KB for classification only; full statements are not exported." That sets expectations and answers a security-review question before it gets asked.
**2. ETL heuristic.** The classifier calls an `INSERT/UPDATE/DELETE/CREATE TABLE "ETL"` if its leading 2KB contains the substring `SELECT`. That's a defensible proxy (`CTAS` / `INSERT ... SELECT` / merge-style updates) but it's not obvious from the SQL itself. A header comment explaining the design would help the next maintainer:
```md
-- Workload classification heuristic:
--   INSERT/UPDATE/DELETE/CREATE TABLE statements that contain a SELECT
--   (within the first 2KB of SQL text) are bucketed as ETL; otherwise
--   they fall into their literal command type. This intentionally
--   prefers false-positive ETL over false-negative — adjust per customer.
```

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/config_containers.sql:
- **-- Has to be executed on CDB** is now enforced by the doc and the connector setup ("must run against CDB$ROOT"), so the inline reminder is redundant. Same comment appears in several other files; if it's kept anywhere, please make it accurate (e.g. "Requires SET CONTAINER to CDB$ROOT" — but honestly, deleting all of them is fine since the doc owns this contract).

Feel free to delete this file if unused.

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/oracle/config_instance.sql:
- **-- Has to be executed on CDB** is now enforced by the doc and the connector setup ("must run against CDB$ROOT"), so the inline reminder is redundant. Same comment appears in several other files; if it's kept anywhere, please make it accurate (e.g. "Requires SET CONTAINER to CDB$ROOT" — but honestly, deleting all of them is fine since the doc owns this contract).
- **-- spool results/config_instance.csv**  is a SQL*Plus output directive that has no effect under `oracledb`. 

Feel free to delete this file if unused.

dgomez04 commented on src/databricks/labs/lakebridge/resources/assessments/validation/oracle_extract_schema.yml:
What is this file for? It's not wired in anywhere in the codebase.

m-abulazm replied:
These are post-run validation schemas for the profiler-dashboard upload step. `create-profiler-dashboard` command runs `_validate_profiler_extract`, which loads the YAML and runs two checks per declared table:                                     
  1. EmptyTableValidationCheck — table contains at least one row.                                                                                                                                                
  2. ExtractSchemaValidationCheck — actual DuckDB column types match the YAML spec.

dgomez04 commented on tests/integration/assessments/test_profiler.py:
- `test_Oracle_*` naming violates **PEP 8** and the surrounding nake_case test function convention. 
- No unit tests for the new `ConfigureOracleAssessment` class, the `OracleConnector` dispatch, or the Oracle branch in `check_connection`. Compare to `legacy_sqldw`, which has all three.

dgomez04 commented on tests/integration/assessments/test_profiler.py:
```diff
@@ -8,6 +8,37 @@
+        profiler.profile()
+
+
+def test_Oracle_profile_execution() -> None:
```
There's no `@pytest.mark.skipif(not env_var)`, no mock or fixture. CI will need an Oracle instance configured or this will fail. Previous pull requests' analogous tests (`tests/unit/...`) used `MockPrompts` and patched the connector; this one doesn't.

dgomez04 commented on tests/integration/assessments/test_profiler.py:
```diff
@@ -8,6 +8,37 @@
+    config_file = path_prefix / "src/databricks/labs/lakebridge/resources/assessments/oracle/pipeline_config.yml"
+    config = profiler.path_modifier(config_file=config_file, path_prefix=path_prefix)
+    profiler.profile(pipeline_config=config)
+    assert Path(
```
Hardcoded `/tmp/...` path for the assertion — same Windows-portability issue. Also: the test will leave a DuckDB file in `/tmp` after running, and there's no cleanup.

dgomez04 commented on docs/lakebridge/docs/assessment/profiler/oracle.mdx:
```diff
@@ -0,0 +1,188 @@
+---
+sidebar_position: 3
```
The new `legacy_sqldw.mdx` (#2441) also uses `sidebar_position: 3`, and `mssql.mdx` already uses `sidebar_position: 2`. If both PRs land, two pages have position 3 and ordering becomes undefined. Whichever lands second needs to bump.

dgomez04 commented on docs/lakebridge/docs/assessment/profiler/oracle.mdx:
```diff
@@ -0,0 +1,188 @@
+
+## Prerequisites
+
+### 1. Required Database Permissions
```
**Numbering** inconsistencies across the section.

dgomez04 commented on docs/lakebridge/docs/assessment/profiler/oracle.mdx:
**No mention of other authentication mechanisms such as TLS / wallet** — Oracle is virtually always behind some authentication boundary in enterprise environments; this gives no hint that wallet/TCPS aren't supported.

**No prerequisite section about the `oracledb` driver.** The other profiler docs explicitly call out the ODBC driver + VC++ runtime; this doc says nothing about Python driver requirements (luckily `oracledb` thin mode needs no native client, so this is informational only — but it's worth saying so).

dgomez04 commented on docs/lakebridge/docs/assessment/profiler/oracle.mdx:
```diff
@@ -0,0 +1,188 @@
+| **Host** | Oracle server hostname, IP, or RAC SCAN name | — |
+| **TNS Listener port** | Port the Oracle listener is bound to | `1521` |
+| **TNS service name** | Service name registered in the Oracle listener (the CDB service) | `orcl` |
+| **User** | Database user with the grants from [Required Database Permissions](#2-required-database-permissions) | `SYSTEM` |
```
The actual heading is `### 1. Required Database Permissions`, so the anchor is `#1-required-database-permissions`. The `#2 ...` anchor is broken.

dgomez04 commented on docs/lakebridge/docs/assessment/profiler/oracle.mdx:
````diff
@@ -0,0 +1,188 @@
+GRANT SELECT ANY DICTIONARY TO c##lakebridge_profiler CONTAINER=ALL;
+```
+
+:::tip
````
**Diagnostic Pack callout is good** — flagging that AWR (`CDB_HIST_*`) requires a paid Oracle add-on. But the YAML's `flag: active` for all `perf_*` steps means you'll silently fail on databases without the license. Should be `flag: inactive` by default for those steps, with the documentation explaining how to flip them on after confirming Diagnostic Pack is licensed.

m-abulazm replied:
I dont expect end users to edit src code. it is called out in the docs and if they fail, that will not break profiler run and querying other tables

dgomez04 commented on docs/lakebridge/docs/assessment/profiler/oracle.mdx:
```diff
@@ -0,0 +1,188 @@
+
+### Profiler Summary Dashboard
+
+Coming Soon
```
Either add or remove this section.

dgomez04 reviewed (changes requested):
Left my comments - let me know if there are any questions or follow-ups I need to address.

dgomez04 commented on .gitignore:
Aren't these just duplicates of the above? Non-blocking but can be cleaned up.

dgomez04 commented on tests/integration/assessments/test_profiler_validator.py:
```diff
@@ -115,7 +115,7 @@ def test_validate_invalid_source_tech(mock_synapse_profiler_extract: Path, test_
             "main",
             "dedicated_routines",
-            source_tech="oracle",
+            source_tech="bogus",
```
Are we not adding the tests with this pull request?

dgomez04 reviewed (changes requested):
Left small comments - please review. 

**Additionally**, can you provide more context on what `/oracle_extract_schema does?`

dgomez04 reviewed (approved):
LGTM - let's keep track of the follow-up issues we opened.

m-abulazm committed:
Add Oracle profiler

Introduces the Oracle source for `configure-database-profiler` and
`execute-database-profiler`. SQLAlchemy + python-oracledb (Thin mode);
CDB-scoped extracts cover configuration metadata (containers, instance,
memory evolution, PDB objects/partitions, storage) and AWR-backed
performance metrics (CPU/waits, foreground session evolution, heatmap,
SQL text).

Includes new docs (docs/lakebridge/docs/assessment/profiler/oracle.mdx),
wiring through the configurator and connector registries, the oracledb
dependency, and integration-test updates.

Co-Authored-By: Laurent Leturgez <laurent.leturgez@databricks.com>
